File: data/src/main/java/com/github/gmetal/data/repository/callback/BaseResponseCallback.java
package com.github.gmetal.data.repository.callback;

import androidx.annotation.NonNull;

import com.github.gmetal.lib.Notifiable;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class BaseResponseCallback<T, R> implements Callback<R> {

    private final ResponseConverter<T, R> mResponseConverter;
    private Notifiable<T, Throwable> mNotifiable;

    public BaseResponseCallback(@NonNull final Notifiable<T, Throwable> notifiable,
                                @NonNull final ResponseConverter<T, R> responseConverter) {
        mNotifiable = notifiable;
        mResponseConverter = responseConverter;
    }

    @Override
    public void onResponse(@NonNull final Call<R> call, @NonNull final Response<R> response) {

        if (response.isSuccessful()) {
            mNotifiable.success(mResponseConverter.convertFrom(response));
        } else {
            mNotifiable.failure(new RuntimeException("Error - Response not successful"));
        }
    }

    @Override
    public void onFailure(@NonNull final Call<R> call, @NonNull final Throwable t) {

        mNotifiable.failure(t);
    }

    public interface ResponseConverter<T, R> {

        T convertFrom(@NonNull final Response<R> response);
    }
}

File: data/src/main/java/com/github/gmetal/data/repository/callback/BaseResponseCallback.kt
package com.github.gmetal.data.repository.callback

import com.github.gmetal.lib.Notifiable
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class BaseResponseCallback<T, R>(private val notifiable: Notifiable<T, Throwable>,
                                 private val responseConverter: (Response<R>) -> T) : Callback<R> {

    override fun onResponse(call: Call<R>, response: Response<R>) {

        if (response.isSuccessful) {
            notifiable.success(responseConverter(response))
        } else {
            notifiable.failure(RuntimeException("Error - Response not successful"))
        }
    }

    override fun onFailure(call: Call<R>, t: Throwable) {

        notifiable.failure(t)
    }
}
